### PR TITLE
Add 0x table on Ethereum in Dune V2

### PR DESCRIPTION
**Pull request by sui414**
Brief comments on the purpose of your changes:

*For Dune Engine V2*
I've checked that:
General checks:
* [x] I tested the query on dune.com after compiling the model with dbt compile (compiled queries are written to the target directory)
* [x] I used "refs" to reference other models in this repo and "sources" to reference raw or decoded tables 
* [x] if adding a new model, I added a test
* [x] the filename is unique and ends with .sql
* [x] each sql file is a select statement and has only one view, table or function defined  
* [x] column names are `lowercase_snake_cased`
* [x] if adding a new model, I edited the dbt project YAML file with new directory path for both models and seeds (if applicable)
* [x] if wanting to expose a model in the UI (Dune data explorer), I added a post-hook in the JINJA config to add metadata (blockchains, sector/project, name and contributors)

Join logic:
* [x] if joining to base table (i.e. ethereum transactions or traces), I looked to make it an inner join if possible

Incremental logic:
* [x] I used is_incremental & not is_incremental jinja block filters on both base tables and decoded tables
* [x] where block_time >= date_trunc("day", now() - interval '1 week')
* [x ] if joining to base table (i.e. ethereum transactions or traces), I applied join condition where block_time >= date_trunc("day", now() - interval '1 week')
* [x] if joining to prices view, I applied join condition where minute >= date_trunc("day", now() - interval '1 week')

### CODE REVIEW

**Comment by sui414** [comment]
didnt find a test template - would love to add it if any link for reference

**Comment by augustog** [comment]
For creating tests, you can follow the pattern used by CoW Protocol [here](https://github.com/duneanalytics/spellbook/pull/1563/files#diff-68d6945ccd0f690131323fe253337945af0c366cb78f6d595dcd6b16112523c3). The query should be defined to match failing rows. 

dbt test will fail if any test returns results.

You could also go the way of dex.trades, where we have a[ generic (parametrised) test](https://github.com/duneanalytics/spellbook/blob/main/tests/generic/dex_trades_token_bought.sql) defined under `tests/generic`, a [seed file with test transactions](https://github.com/duneanalytics/spellbook/blob/main/seeds/dex/trades/dex_trades_seed.csv), and [every spell defines the test in its schema file](https://github.com/duneanalytics/spellbook/blob/f097fe1898fd5cf45bd479fffdf95466c1e48bd3/models/curvefi/ethereum/curvefi_ethereum_schema.yml#L169). 

I would suggest that approach to DRY if you envision multiple spells with the same / very similar schema.

**Comment by Hosuke** [comment]
Hi @sui414 @augustog ,
I think I could help with adding tests in this PR.
I will make up a seed file with a few records in 0x table (and verify on etherscan).
Then I will add some generic tests in the schema file to test for some specific column value matches.

File: models/zeroex/ethereum/zeroex_api_ethereum_schema.yml
Comment (Hosuke) [inline models/zeroex/ethereum/zeroex_api_ethereum_schema.yml:65]: You may missed a column `volume_usd` here.

File: models/zeroex/ethereum/zeroex_api_ethereum_schema.yml
Comment (Hosuke) [inline models/zeroex/ethereum/zeroex_api_ethereum_schema.yml:12]: You can add an unique keys testing here (after the `description`, and before the `columns`)
The test code looks like
```
    tests:
      - dbt_utils.unique_combination_of_columns:
          combination_of_columns:
            - tx_hash
            - evt_index
```

File: models/zeroex/ethereum/zeroex_api_ethereum_schema.yml
Comment (Hosuke) [inline models/zeroex/ethereum/zeroex_api_ethereum_schema.yml:41]: After the test seeds added, you can test the column value by adding lines:
```
      - &maker_token
        name: maker_token
        description: "Contract address of the token taker buys"
        tests:
          - dex_trades_token_bought:
              dex_trades_seed: ref('zeroex_ethereum_transactions')
```

**Review by Hosuke — CHANGES_REQUESTED** [review]
For adding tests:

1.  you can create a directory under `seeds`, like `seeds/zeroex/ethereum`
2. then you can add a seed file in this directory, eg. `seeds/zeroex/ethereum/zeroex_ethereum_transactions.csv`
3. And paste some verified records in this file, with csv format. eg. 
```
tx_hash,evt_index,contract_address,block_time,maker,taker,taker_token,maker_token,taker_token_amount,taker_token_amount_raw,maker_token_amount,maker_token_amount_raw,type,affiliate_address,swap_flag,matcha_limit_order_flag,volume_usd
0x55081d325534be2f97cf1a49702787f47bd08ac2e25a81fc2a4445754209f76b,16,0x7b504a15ef05f4eed1c07208c5815c49022a0c19,2022-10-27 02:09,0x7b504a15ef05f4eed1c07208c5815c49022a0c19,0xe66b31678d6c16e9ebf358268a790b763c133750,0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2,0xc944e90c64b2c07662a292be6244bdf05cda44a7,,1.90119544865326E+16,357.230944900262,3.57230944900262E+20,Sushiswap Direct,0x382ffce2287252f930e1c8dc9328dac5bf282ba1,TRUE,FALSE,29.7273303108202
0x14493b0f901511657f5ff793241c780175707f209b037a582127c51a917a8743,114,0x7b73644935b8e68019ac6356c40661e1bc315860,2022-10-27 02:40,0x7b73644935b8e68019ac6356c40661e1bc315860,0xf7ae03ebc56d587b89de0e54b59f588712a7961e,0x761d38e5ddf6ccf6cf7c55759d5210750b5d60f3,0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2,,2.17289415972178E+26,,3.80612498664961E+16,Uniswap V2 Direct,0x382ffce2287252f930e1c8dc9328dac5bf282ba1,TRUE,FALSE,
0x74be02cb68a3b181664fccfc1d1bea48038e7f0f28c59b9f8c273aa00e28d8f4,125,0x4bb0925fa50da9b4c8936869433b48e78ccc5c13,2022-10-27 02:07,0x4bb0925fa50da9b4c8936869433b48e78ccc5c13,0xe66b31678d6c16e9ebf358268a790b763c133750,0xdac17f958d2ee523a2206206994597c13d831ec7,0x5b7533812759b45c2b44c19e320ba2cd2681b542,,9.8665962E+07,,1.83794614553E+11,Uniswap V2 Direct,0x382ffce2287252f930e1c8dc9328dac5bf282ba1,TRUE,FALSE,
0x4220ff45f296b5ab809e11574911c2eb4644502f1496b0695f26dcddd00377be,4,0xa3509a16bbfc5992eb01cc861b615ccd8e937da8,2022-10-27 02:12,0xa3509a16bbfc5992eb01cc861b615ccd8e937da8,0x74de5d4fcbf63e00296fd95d33236b9794016631,0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2,0xde30da39c46104798bb5aa3fe8b9e0e1f348163f,,1.9825E+16,,1.85397121437985E+19,Uniswap V2 Direct,0x11ededebf63bef0ea2d2d071bdf88f71543ec6fb,TRUE,FALSE,
0x440c41132c739a05898a2b428a67ffe26dc78a4899c391cfcaa52707c5498557,8,0x9c4cc862f51b1ba90485de3502aa058ca4331f32,2022-10-27 02:02,0x9c4cc862f51b1ba90485de3502aa058ca4331f32,0x74de5d4fcbf63e00296fd95d33236b9794016631,0xc02aaa39b223fe8d0a0e5c4f27ead9083c756cc2,0x26c8afbbfe1ebaca03c2bb082e69d0476bffe099,,3.965E+17,2680.5228765013,2.6805228765013E+21,Uniswap V2 Direct,0x11ededebf63bef0ea2d2d071bdf88f71543ec6fb,TRUE,FALSE,603.795819500547

```
4. Then you make this seed available by adding seeds to `dbt_project.yml`, under `seeds:`

eg.
```
    zeroex:
      ethereum:
        +enabled: true
        +schema: test_data
        zeroex_ethereum_transactions:
          +column_types:
            tx_hash: string
            evt_index: int
            contract_address: string
            block_time: timestamp
            block_date: date
            maker: string
            taker: string
            taker_token: string
            maker_token: string
            taker_token_amount: double
            taker_token_amount_raw: double
            maker_token_amount: double
            maker_token_amount_raw: double
            type: string
            affiliate_address: string
            swap_flag: bool
            matcha_limit_order_flag: bool
            volume_usd: double
```

Hope these information can be helpful for adding tests.

**Comment by Hosuke** [comment]
Hi @sui414, if you are new to dbt stuff or busy with other stuff, you can add me as a collaborator in your forked repo. I will push commits to the forked repo and help with 0x spells before they get merged.

**Comment by sui414** [comment]
> Hi @sui414, if you are new to dbt stuff or busy with other stuff, you can add me as a collaborator in your forked repo. I will push commits to the forked repo and help with 0x spells before they get merged.

thank you so much for the help on review and tips!

File: models/zeroex/ethereum/zeroex_api_ethereum_fills.sql
Comment (Hosuke) [inline models/zeroex/ethereum/zeroex_api_ethereum_fills.sql:17]: Hi @sui414, I have added the project_start_date for zeroex v3 and v4, these dates will be used in full-refresh for performance gain. There are some CTEs I am not sure which date should I use, so I use the default v3 date.

These dates are based on [the query](https://dune.com/queries/1559388).

Comment (sui414) [reply models/zeroex/ethereum/zeroex_api_ethereum_fills.sql:17]: thank you! seems accurate to me. I added you as a collaborator for this PR. 
gonna finish solving the test error on unique id and ask dune team for Merge soonish.

Comment (sui414) [reply models/zeroex/ethereum/zeroex_api_ethereum_fills.sql:17]: updated some of the subqueries' dates

**Comment by sui414** [comment]
@augustog @Hosuke fixed the test errors - plz let me know if there is any other issue blocking the merge!

**Comment by Hosuke** [comment]
Thanks for the update @sui414 , it's now passing the tests. I will try to tidy up the codes today, and pass the PR to Jeff for a final review for merging.

**Comment by Hosuke** [comment]
I tried to reuse spells/views in sushiswap and uniswap, however it break the code(failed in uniqueness test), so I revert those changes.

**Comment by Hosuke** [comment]
To whom it may concern, I have reviewed this PR and here is the summary:

- Metadata and naming convention looks good ✅
- Config blocks match uniqueness tests in schema files ✅
- Tests and seeds added for the new models ✅
- Incremental filters added when joining large tables ✅
- Contract addresses are lowercased ✅
- Inner join to large tables(ethereum.transactions) as possible. However there is heavy use of logs and traces table.

I tried to enhance the model performance by using ref('sushiswap_ethereum_trades'), ref('uniswap_v2..') and ref('uniswap_v3..') but failed. The model logics are complex and decodes [logs, traces] on the run.

Possible improvements in future? :
- break down into smaller models and aggregate in meta table (which is easier for maintenance)
- try leveraging existing spells/views in the union all section

File: models/zeroex/ethereum/zeroex_api_ethereum_schema.yml
Comment (jeff-dude) [reply models/zeroex/ethereum/zeroex_api_ethereum_schema.yml:47]: i don't think it is necessary

**Review by jeff-dude — APPROVED** [review]
phew, fantastic spell PR @sui414 🙌
and thanks for the support, @Hosuke!

**Pull request merged**
